File: core/src/main/java/com/sequenceiq/cloudbreak/controller/mapper/DuplicatedKeyValueExceptionMapper.java
package com.sequenceiq.cloudbreak.controller.mapper;

import static ch.qos.logback.classic.Level.INFO;

import javax.ws.rs.core.Response.Status;

import org.springframework.stereotype.Component;

import com.sequenceiq.cloudbreak.exception.mapper.BaseExceptionMapper;
import com.sequenceiq.cloudbreak.service.DuplicateKeyValueException;

import ch.qos.logback.classic.Level;

@Component
public class DuplicatedKeyValueExceptionMapper extends BaseExceptionMapper<DuplicateKeyValueException> {

    @Override
    protected Level getLogLevel() {
        return INFO;
    }

    @Override
    public Status getResponseStatus(DuplicateKeyValueException exception) {
        return Status.CONFLICT;
    }

    @Override
    public Class<DuplicateKeyValueException> getExceptionType() {
        return DuplicateKeyValueException.class;
    }

    public static String errorMessage(DuplicateKeyValueException exception) {
        return String.format("The %s name '%s' is already taken, please choose a different one",
                exception.getResourceType().toString().toLowerCase(),
                exception.getValue());
    }
}

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/mapper/DuplicatedKeyValueExceptionMapper.kt
package com.sequenceiq.cloudbreak.controller.mapper

import javax.ws.rs.core.Response
import javax.ws.rs.ext.ExceptionMapper
import javax.ws.rs.ext.Provider

import org.slf4j.Logger
import org.slf4j.LoggerFactory

import com.sequenceiq.cloudbreak.controller.json.ExceptionResult
import com.sequenceiq.cloudbreak.logger.MDCBuilder
import com.sequenceiq.cloudbreak.service.DuplicateKeyValueException

@Provider
class DuplicatedKeyValueExceptionMapper : ExceptionMapper<DuplicateKeyValueException> {

    override fun toResponse(exception: DuplicateKeyValueException): Response {
        MDCBuilder.buildMdcContext()
        LOGGER.error(exception.message, exception)
        return Response.status(Response.Status.CONFLICT).entity(ExceptionResult(
                String.format("The %s name '%s' is already taken, please choose a different one",
                        exception.resourceType.toString().toLowerCase(),
                        exception.value))).build()
    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger(DuplicatedKeyValueExceptionMapper::class.java)
    }
}
